### PR TITLE
Fix DeepSeek V4 sliding window attention mask for packed sequences

### PR DESCRIPTION
**Pull request by snehalv2002**
# Description

Fixes multiple edge cases in `AttentionOp.generate_attention_mask` for `AttentionType.COMPRESSED` (DeepSeek-V4 attention) across chunked prefill, packed sequences, and standalone test environments.

### Context & Fixes
1. **Chunked Prefill & Prompt Caching ($Q < S$)**:
   - On `main`, `abs_k = segment_positions[:, None, :s_len]` slices `segment_positions`. In chunked prefill where $q\_seq\_len < s\_len$ (e.g. $Q=4, S=8$), `segment_positions` only contains $Q$ elements, causing an invalid $(B, 1, Q)$ slice instead of $S$ columns, resulting in a broadcasting shape mismatch (`ValueError: Incompatible types for broadcasting: input type=bool[1,1,1,4,4] and requested type=bool[1,1,1,4,8]`).
   - **Fix**: For standard prefill and training, uses contiguous global buffer coordinates `row_ids = [0..Q-1] + next_pos` and `col_ids = [0..S-1]`, guaranteeing proper shape broadcasting across all query chunk sizes.
2. **Autoregressive Decode without Segment IDs (`decoder_segment_ids=None`)**:
   - Fixes sliding window distance calculations during single-token AR decode ($Q=1$) when segment IDs are not provided by falling back to `next_pos` and buffer coordinates.
3. **`mesh=None` Guard in `_maybe_shard_with_pspec`**:
   - Added `if inputs is None or self.mesh is None or pspec is None: return inputs` to prevent `TypeError: NamedSharding(None, P)` crashes in unit tests and single-device execution where no device mesh is active.
4. **Packed Sequence Document Isolation**:
   - Ensures causal/sliding distance is valid across concatenated documents in packed datasets (e.g. C4) without intra-document attention collapse, while document boundaries are strictly and cleanly isolated by `decoder_segment_ids`.
5. **Code Documentation**:
   - Added comprehensive docstrings and case-by-case inline comments explaining the uncompressed sliding window prefix and compressed memory block attention math for reviewers.

# Tests
- Added `test_generate_attention_mask_compressed_chunked_prefill` in `tests/unit/attention_test.py` (verifies $Q=4, S=8$ chunked prefill with `previous_chunk` and `segment_positions`; reproduces failure on `main` and passes with this PR).
- Added `test_generate_attention_mask_compressed_ar_without_segment_ids` in `tests/unit/attention_test.py` (verifies $Q=1$ AR decode with `decoder_segment_ids=None`).
- Added `test_generate_attention_mask_packed_sequences` in `tests/unit/attention_test.py` (verifies multi-document packed sequences).
- Added `test_generate_attention_mask_compressed_all_modes` in `tests/unit/attention_test.py` (verifies train, prefill, and AR modes).
- Added `test_generate_attention_mask_compressed_load_balanced_context_parallel` in `tests/unit/attention_test.py` (verifies `segment_positions` support under load-balanced CP).
- Verified full test suite passes on TPU v5p-8 (`15/15 tests PASSED`, including Flash vs Dot-Product parity tests in `CompressedAttentionTest`).
- **End-to-End Verification Run**: TPU v5p-128 cluster `mlperf-v5p` (Step 0 `lm_loss`: **2.823**, healthy baseline).
  - [Cloud Logging Link](https://console.cloud.google.com/logs/query;query=resource.type%3D%22k8s_container%22%0Aresource.labels.project_id%3D%22cloud-tpu-multipod-dev%22%0Aresource.labels.location%3D%22europe-west4%22%0Aresource.labels.cluster_name%3D%22mlperf-v5p%22%0Aresource.labels.namespace_name%3D%22default%22%0Aresource.labels.pod_name%3A%22dsv4-fix-verify-1787781315-slice-job-0-0-%22%0Aseverity%3E%3DDEFAULT;storageScope=project;duration=P1D?project=cloud-tpu-multipod-dev)
  - [Launch Command Paste](https://paste.googleplex.com/4883586289893376)

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [x] I have performed a self-review of my code. For an optional AI review, add the `gemini-review` label.
- [x] I have necessary comments in my code, particularly in hard-to-understand areas.
- [x] I have run end-to-end tests and provided workload links above if applicable.
- [x] I have made or will make corresponding changes to the doc if needed.

### CODE REVIEW

**Review by shuningjin — APPROVED** [review]

File: tests/unit/attention_test.py [hunk 1]
```diff
@@ -4772,6 +4871,64 @@ def test_generate_attention_mask_compressed_all_modes(self):
     )
     self.assertEqual(mask_none.ndim, 4)
     self.assertEqual(mask_none.shape[-1], kv_len)
+    mask_none_np = np.array(mask_none)[0, 0]
+    self.assertEqual(mask_none_np[0, 0], 0.0)
+    self.assertEqual(mask_none_np[0, 1], DEFAULT_MASK_VALUE)
+
+  def test_generate_attention_mask_compressed_load_balanced_context_parallel(self):
```
Comment (octatrifan) [inline tests/unit/attention_test.py:4991]: I've tried to run these updated tests `test_generate_attention_mask_compressed_all_modes` and `test_generate_attention_mask_compressed_load_balanced_context_parallel` and they seem to already pass with the code on the main branch. Could you please run them to confirm this?

File: tests/unit/attention_test.py [hunk 2]
```diff
@@ -4604,7 +4604,12 @@ class DeepSeekV4AttentionMaskingTest(unittest.TestCase):
   """Tests to validate AttentionOp masking logic for DeepSeek-V4 attention patterns."""
```
Comment (octatrifan) [inline tests/unit/attention_test.py:4604]: Some tests that might break on main but work with this PR are:
- Chunked prefill (`q_len < s_len` with `segment_positions`)
- AR decode without segment IDs (`decoder_segment_ids=None`)

Could be useful to add them and check that this PR indeed fixes these scenarios.

**Comment by octatrifan** [comment]
> Previously, for training and prefill phases (q_seq_len > 1), abs_q was derived from segment_positions[:, :, None], while abs_k used global buffer indices jnp.arange(s_len)[None, None, :].

PR [#4488](https://github.com/AI-Hypercomputer/maxtext/pull/4488) already patched `abs_k` on main to use `segment_positions[:, None, :s_len]`. So that specific asymmetry should not be a problem.

That said, the PR still fixes the following:
1. `_maybe_shard_with_pspec` crashed whenever `mesh is None` (the tests in this PR catch this)
2. On main, `abs_k = segment_positions[:, None, :s_len]` breaks if `q_seq_len < s_len` (chunked prefill / prompt caching) 
3. The comments added are great for code readability.

Could you please review the description and the current main to see if this is true?

File: tests/unit/attention_test.py (hunk 2 above)
Comment (snehalv2002) [reply tests/unit/attention_test.py:4604]: Added these tests and verified that they break on main but pass here 👍

File: tests/unit/attention_test.py (hunk 1 above)
Comment (snehalv2002) [reply tests/unit/attention_test.py:4991]: yes you're right they do pass on `main` because in in that test, `Q = S` (full prefill), so `segment_positions` has the same length as the uncompressed prefix (`S`), which avoids the slicing shape mismatch that occurs when `Q < S` (chunked prefill). i've added the tests you mentioned below which show errors on main but succeed here 👍

**Review by octatrifan — APPROVED** [review]
